// camunda-8-credentials-from-env/dist/lib/tasklist.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports._getTasklistEnv = void 0;
const neon_env_1 = require("neon-env");
function _getTasklistEnv() {
    return (0, neon_env_1.createEnv)({
        ZEEBE_ADDRESS: {
            type: 'string',
            optional: false
        },
        ZEEBE_CLIENT_ID: {
            type: 'string',
            optional: false
        },
        ZEEBE_CLIENT_SECRET: {
            type: 'string',
            optional: false
        },
        ZEEBE_AUTHORIZATION_SERVER_URL: {
            type: 'string',
            optional: false
        },
        ZEEBE_TOKEN_AUDIENCE: {
            type: 'string',
            optional: true
        },
        CAMUNDA_CLUSTER_ID: {
            type: 'string',
            optional: false
        },
        CAMUNDA_CLUSTER_REGION: {
            type: 'string',
            optional: false
        },
        CAMUNDA_CREDENTIALS_SCOPES: {
            type: 'string',
            optional: false
        },
        CAMUNDA_TASKLIST_BASE_URL: {
            type: 'string',
            optional: true
        },
        CAMUNDA_OAUTH_URL: {
            type: 'string',
            optional: false
        }
    });
}
exports._getTasklistEnv = _getTasklistEnv;
