// optimize-api-client/dist/lib/ReportResults.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.ReportResults = void 0;
const got_1 = __importDefault(require("got"));
class ReportResults {
    constructor({ getHeaders, gotOptions, limit, reportId, paginationTimeout, }) {
        this.getHeaders = getHeaders;
        this.gotOptions = gotOptions;
        this.reportId = reportId;
        this.limit = limit;
        this.paginationTimeout = paginationTimeout;
    }
    next() {
        return this.getResultsPage();
    }
    getResultsPage(searchRequestId) {
        return __awaiter(this, void 0, void 0, function* () {
            const headers = yield this.getHeaders();
            const sreqId = !!this.searchRequestId
                ? `&searchRequestId=${this.searchRequestId}`
                : ``;
            const result = yield (0, got_1.default)(`export/report/${this.reportId}/result/json?paginationTimeout=${this.paginationTimeout}&limit=${this.limit}${sreqId}`, Object.assign({ headers }, this.gotOptions)).json();
            this.searchRequestId = result.searchRequestId;
            return result;
        });
    }
}
exports.ReportResults = ReportResults;
