// camunda-saas-oauth/dist/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.getConsoleToken = exports.getZeebeToken = exports.getTasklistToken = exports.getOptimizeToken = exports.getOperateToken = exports.OAuthProvider = exports.OAuthProviderImpl = void 0;
var OAuthProviderImpl_1 = require("./lib/OAuthProviderImpl");
Object.defineProperty(exports, "OAuthProviderImpl", { enumerable: true, get: function () { return OAuthProviderImpl_1.OAuthProviderImpl; } });
var OAuthProvider_1 = require("./lib/OAuthProvider");
Object.defineProperty(exports, "OAuthProvider", { enumerable: true, get: function () { return OAuthProvider_1.OAuthProvider; } });
var Operate_1 = require("./lib/Operate");
Object.defineProperty(exports, "getOperateToken", { enumerable: true, get: function () { return Operate_1.getOperateToken; } });
var Optimize_1 = require("./lib/Optimize");
Object.defineProperty(exports, "getOptimizeToken", { enumerable: true, get: function () { return Optimize_1.getOptimizeToken; } });
var Tasklist_1 = require("./lib/Tasklist");
Object.defineProperty(exports, "getTasklistToken", { enumerable: true, get: function () { return Tasklist_1.getTasklistToken; } });
var Zeebe_1 = require("./lib/Zeebe");
Object.defineProperty(exports, "getZeebeToken", { enumerable: true, get: function () { return Zeebe_1.getZeebeToken; } });
var Console_1 = require("./lib/Console");
Object.defineProperty(exports, "getConsoleToken", { enumerable: true, get: function () { return Console_1.getConsoleToken; } });
